src/riak_kv_backend.erl

```diff
@@ -31,33 +30,52 @@
 -type fold_objects_fun() :: fun((binary(), binary(), term(), any()) ->
                                        any() |
                                        no_return()).
+
+-type index_spec() :: {add, Index, SecondaryKey} | {remove, Index, SecondaryKey}.
```

Comment:
After discussing with @nickelization the spec we arrived at the conclusion that the spec 

`
-type index_spec() :: {add, Index, SecondaryKey} | {remove, Index, SecondaryKey}.
`

may not be functionally identical to

`-type index_spec() :: {add, Index :: any(), SecondaryKey :: any()} | {remove, Index :: any(), SecondaryKey :: any()}.`

And is likely a polymorphic type declaration interpreted as, a union of two three tuples where
 - the first element is the atom `add` or `remove`
 - the second element must have the same type
 - the third element must have the same type 

Example
- VALID `{add, one, <<1>>}` and `{remove, one, <<2>>}` 
- INVALID `{add, [one], <<>>}` and `{remove, {one}, <<>>}`
- ????  `{add, {one, 1}, <<1>>}` and `{remove, {one, 2}, <<2>>}`